Flutter_TDD_Architecture_Course_notes.md

```diff
@@ -277,3 +277,740 @@ class GetRandomNumberTrivia extends UseCase<NumberTrivia, NoParams> {
+```dart
+class ServerException implements Exception {}
+
+class CacheException implements Exception {}
+```
+This is the code for the `failure.dart` file that also has to be created inside the `error` folder.
+
+```dart
+import 'package:equatable/equatable.dart';
+
+abstract class Failure extends Equatable {
+  Failure([List properties = const <dynamic>[]]) : super(properties);
+}
+
+// General failures
+class ServerFailure extends Failure {}
+
+class CacheFailure extends Failure {}
+```
+Now we need to create a file for the `NumberTriviaRepositoryImpl` tests.
+This file will be named `number_trivia_repository_impl_test.dart`.
+Then we prepare the file for the test.
+
+```dart
+class MockRemoteDataSource extends Mock
+    implements NumberTriviaRemoteDataSource {}
+
+class MockLocalDataSource extends Mock implements NumberTriviaLocalDataSource {}
+
+class MockNetworkInfo extends Mock implements NetworkInfo {}
+
+void main() {
+  NumberTriviaRepositoryImpl repository;
+  MockRemoteDataSource mockRemoteDataSource;
+  MockLocalDataSource mockLocalDataSource;
+  MockNetworkInfo mockNetworkInfo;
+
+  setUp(() {
+    mockRemoteDataSource = MockRemoteDataSource();
+    mockLocalDataSource = MockLocalDataSource();
+    mockNetworkInfo = MockNetworkInfo();
+    repository = NumberTriviaRepositoryImpl(
+      remoteDataSource: mockRemoteDataSource,
+      localDataSource: mockLocalDataSource,
+      networkInfo: mockNetworkInfo,
+    );
+  });
+}
+```
+If we run the test now it will give an error because have not yet written the implementation. 
+Open the `number_trivia_repository_impl.dart` file and add the following code:
+
+```dart
+import 'package:dartz/dartz.dart';
+import 'package:meta/meta.dart';
+
+import '../../../../core/error/failure.dart';
+import '../../../../core/platform/network_info.dart';
+import '../../domain/entities/number_trivia.dart';
+import '../../domain/repositories/number_trivia_repository.dart';
+import '../datasources/number_trivia_local_data_source.dart';
+import '../datasources/number_trivia_remote_data_source.dart';
+
+class NumberTriviaRepositoryImpl implements NumberTriviaRepository {
+  final NumberTriviaRemoteDataSource remoteDataSource;
+  final NumberTriviaLocalDataSource localDataSource;
+  final NetworkInfo networkInfo;
+
+  NumberTriviaRepositoryImpl({
+    @required this.remoteDataSource,
+    @required this.localDataSource,
+    @required this.networkInfo,
+  });
+
+  @override
+  Future<Either<Failure, NumberTrivia>> getConcreteNumberTrivia(int number) {
+    // TODO: implement getConcreteNumberTrivia
+    return null;
+  }
+
+  @override
+  Future<Either<Failure, NumberTrivia>> getRandomNumberTrivia() {
+    // TODO: implement getRandomNumberTrivia
+    return null;
+  }
+}
+```
+#### 6. Repository Implementation
+> Video 6: https://www.youtube.com/watch?v=bfEKPKKy9dA
+
+In the `sixth` tutorial start by explaining that we will have to know if the user has an internet connection and for that a test will be done.
```

Comment:
Grammar and excess word count. 
![image](https://user-images.githubusercontent.com/194400/75149177-4f917680-56f9-11ea-801c-28c516837481.png)

```suggestion
The `sixth` video starts by explaining that we will have to know if the user has an internet connection.
```